File: servers/nodejs/https-server.js
import { createServer } from "https"
import { readFileSync } from "fs"

// Read the SSL certificate files
const privateKey = readFileSync("./ssl/key.pem", "utf8")
const certificate = readFileSync("./ssl/cert.pem", "utf8")

const defaultPort = 3000

// Get the port from the command line arguments
const port = process.argv[2] || defaultPort

const server = createServer(
  {
    key: privateKey,
    cert: certificate,
    minVersion: "TLSv1.3",
    maxVersion: "TLSv1.3",
  },
  (req, res) => {
    res.writeHead(200, { "Content-type": "text/plain" })
    res.end("1")
  }
)

// Using exclusive: false allows port to be shared and reused
server.listen(port, {
  exclusive: false,
})

File: servers/ruby/https-server.rb
require 'webrick'
require 'webrick/https'

default_port = 3000

# Get the port from the command line arguments
port = ARGV[0] ? ARGV[0].to_i : default_port

server = WEBrick::HTTPServer.new(
  Port: port,
  Logger: WEBrick::Log::new("/dev/null"),   # Disable general logging
  AccessLog: [],                            # Disable access logging
  SSLEnable: true,
  SSLCertificate: OpenSSL::X509::Certificate.new(File.open(File.join("ssl", "cert.pem")).read),
  SSLPrivateKey: OpenSSL::PKey::RSA.new(File.open(File.join("ssl", "key.pem")).read),
  SSLVersion: 'TLSv1_3',                     # Specify to use only TLS 1.3
  StartCallback: proc {
    server.listeners.each do |s|
      s.setsockopt(:SOCKET, :REUSEPORT, 1)
    end
  }
)

server.mount_proc '/' do |req, res|
  res.body = '1'
  res['Content-Type'] = 'text/plain'
end

# Handle signals for graceful shutdown
trap 'INT' do server.shutdown end
trap 'TERM' do server.shutdown end

server.start
